### ISSUE TITLE
[BUG] Broadcasting multi-qubit gates in tape mode produces new duplicate wires

### ISSUE THREAD

**Issue by glassnotes** [PennyLaneAI/pennylane#1030]
#### Issue description

When using the broadcast function in tape mode, broadcasting multi-qubit gates seems to create new wires that have type tensor.

* *Expected behavior:* Broadcasted gates should be applied on the device wires.

* *Actual behavior:* Additional wires are created that have type tensor.

* *Reproduces how often:* Whenever multi-qubit gates are included in the broadcast. 

* *System information:* 
(Using dev version from 2020-01-24)

```
Name: PennyLane
Version: 0.14.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /home/olivia/Software/anaconda3/envs/xanadu/lib/python3.8/site-packages
Requires: scipy, networkx, semantic-version, autograd, toml, numpy, appdirs
Required-by: PennyLane-SF, pennylane-qulacs, PennyLane-qsharp, PennyLane-qiskit, PennyLane-Qchem, PennyLane-Forest, PennyLane-Cirq
Platform info:           Linux-5.4.0-64-generic-x86_64-with-glibc2.10
Python version:          3.8.5
Numpy version:           1.19.5
Scipy version:           1.4.1
Installed devices:
- default.gaussian (PennyLane-0.14.0.dev0)
- default.mixed (PennyLane-0.14.0.dev0)
- default.qubit (PennyLane-0.14.0.dev0)
- default.qubit.autograd (PennyLane-0.14.0.dev0)
- default.qubit.jax (PennyLane-0.14.0.dev0)
- default.qubit.tf (PennyLane-0.14.0.dev0)
- default.tensor (PennyLane-0.14.0.dev0)
- default.tensor.tf (PennyLane-0.14.0.dev0)
- strawberryfields.fock (PennyLane-SF-0.11.0)
- strawberryfields.gaussian (PennyLane-SF-0.11.0)
- strawberryfields.remote (PennyLane-SF-0.11.0)
- qulacs.simulator (pennylane-qulacs-0.12.0)
- microsoft.QuantumSimulator (PennyLane-qsharp-0.8.0)
- qiskit.aer (PennyLane-qiskit-0.13.0.dev0)
- qiskit.basicaer (PennyLane-qiskit-0.13.0.dev0)
- qiskit.ibmq (PennyLane-qiskit-0.13.0.dev0)
- forest.numpy_wavefunction (PennyLane-Forest-0.11.0)
- forest.qpu (PennyLane-Forest-0.11.0)
- forest.qvm (PennyLane-Forest-0.11.0)
- forest.wavefunction (PennyLane-Forest-0.11.0)
- cirq.mixedsimulator (PennyLane-Cirq-0.13.0)
- cirq.pasqal (PennyLane-Cirq-0.13.0)
- cirq.qsim (PennyLane-Cirq-0.13.0)
- cirq.qsimh (PennyLane-Cirq-0.13.0)
- cirq.simulator (PennyLane-Cirq-0.13.0)


```

#### Source code and tracebacks

```
import pennylane as qml
import numpy as np
qml.enable_tape()

N_states = 4
params = np.random.uniform(low=-np.pi/2, high=np.pi/2, size=(N_states))

def circuit(params, wires):
    qml.broadcast(qml.RY, wires, pattern='single', parameters=params)
    return qml.expval(qml.PauliZ(0))

dev = qml.device("default.qubit", wires=N_states)
my_qnode = qml.QNode(circuit, dev)
my_qnode(params, range(N_states))
print(my_qnode.draw())
```
This produces the expected output:

```
 0: ──RY(-0.195)──┤ ⟨Z⟩ 
 1: ──RY(-0.18)───┤     
 2: ──RY(-0.779)──┤     
 3: ──RY(0.404)───┤     
```

However altering the function to include a multi-qubit gate, like so:
```
def circuit(params, wires):
    qml.broadcast(qml.RY, wires, pattern='single', parameters=params)
    qml.broadcast(qml.CRY, wires, pattern='ring', parameters=params)
    return qml.expval(qml.PauliZ(0))
```
gives the following output after running and drawing the tape:
```
 0: ──RY(-0.21)──╭C─────────────────────────────────╭RY(1.14)──┤ ⟨Z⟩ 
 1: ──RY(1.38)───╰RY(-0.21)──╭C─────────────────────│──────────┤     
 2: ──RY(-1.05)──────────────╰RY(1.38)──╭C──────────│──────────┤     
 3: ──RY(1.14)──────────────────────────╰RY(-1.05)──╰C─────────┤     
 0: ──RY(-0.21)───C──────────────────────────────────RY(1.14)──┤ ⟨Z⟩ 
 1: ──RY(1.38)────RY(-0.21)───C────────────────────────────────┤     
 2: ──RY(-1.05)───────────────RY(1.38)───C─────────────────────┤     
 3: ──RY(1.14)───────────────────────────RY(-1.05)───C─────────┤     

```
It doesn't seem to be an issue just in the tape drawer, because looking at the tape operations shows that the CRY are applied to different wires:
```
[RY(tensor(-0.20956558, requires_grad=True), wires=[0]),
 RY(tensor(1.37704322, requires_grad=True), wires=[1]),
 RY(tensor(-1.05444866, requires_grad=True), wires=[2]),
 RY(tensor(1.14254618, requires_grad=True), wires=[3]),
 CRY(tensor(-0.20956558, requires_grad=True), wires=[tensor(0, requires_grad=True), tensor(1, requires_grad=True)]),
 CRY(tensor(1.37704322, requires_grad=True), wires=[tensor(1, requires_grad=True), tensor(2, requires_grad=True)]),
 CRY(tensor(-1.05444866, requires_grad=True), wires=[tensor(2, requires_grad=True), tensor(3, requires_grad=True)]),
 CRY(tensor(1.14254618, requires_grad=True), wires=[tensor(3, requires_grad=True), tensor(0, requires_grad=True)])]
```

**Comment by josh146** [comment]
woah, thanks for catching this @glassnotes.

This is a consequence of two things:

* Currently, all QNode parameters are assumed to be differentiable by default when using Autograd. This is to ensure backwards compatibility with old demos, where we pass floats/lists to QNodes and optimize them. At some point soon, we will switch to assuming `requires_grad=False` **by default** for all floats/lists/arrays (`requires_grad=True` will need to be explicit, like in PyTorch), which will avoid issues like this.

* Further, because we currently assume `requires_grad=True` by default, all input arguments become `qml.np.array` objects. These arrays have slightly different indexing behaviour to vanilla NumPy arrays.

  ```python
  >>> x = np.array([0, 1, 2], requires_grad=True)
  >>> x[0]
  tensor(0, requires_grad=True)
  ```

I think the current best fix is to take this into account in the `wires.py` file, like so:

```python
def _process(wires):
    """Converts the input to a tuple of numbers or strings."""

    if isinstance(wires, (Number, str)):
        # interpret as a single wire
        return (wires,)

    if isinstance(wires, Wires):
        # if input is already a Wires object, just return its wire tuple
        return wires.labels

    if getattr(wires, "shape", None) == tuple():
        # Scalar NumPy array
        return (wires.item(),)

    # New if statement here
    if isinstance(wires, np.ndarray):
        wires = wires.view(np.ndarray)

    if isinstance(wires, Iterable):
```

The new `if` statement checks if the provided wires are a subclass of `ndarray`, and if so, casts it to a vanilla NumPy array.

**Comment by mariaschuld** [comment]
While @josh146 's hotfix should solve the particular problem, I think we could address it on a deeper level (also to avoid bloating the `_process` function.

For context: what we removed in PR #982 is some complicated preprocessing of wire labels - now a wire is identified by a unique object in python (i.e. two objects for which `__eq__` is true define the same wire). This is a clean logic, and if both broadcast functions would do the same conversion to an array, everything should still work fine. 

So what is wrong here is that broadcast acting on a 2-qubit "pattern" casts the wires, while broadcast on 1 wire does not.

I will look into it!

**Comment by mariaschuld** [comment]
I dug quite deep into the issue and I think there are two of them, one of which is easy to fix.

The first issue is that broadcast with single pattern has a different behaviour to broadcast with double pattern. This is a classic bug, and can be fixed by changing
``` python
 "single": lambda wires: [wires[i] for i in range(len(wires))],
 "double": lambda wires: [wires.subset([i, i + 1]) for i in range(0, len(wires) - 1, 2)],
```
to 
``` python
 "single": lambda wires: [wires.subset([i]) for i in range(len(wires))],
 "double": lambda wires: [wires.subset([i, i + 1]) for i in range(0, len(wires) - 1, 2)],
```
so that they both follow the same logic. (This did not disturb us before because indexing returned a wires object, so they did the same.) 

The second issue has nothing to do with broadcast but with a combination of `pennylane.numpy.array` indexing and the conversion of inputs to this data structure for backwards compatibility:

```python
from pennylane import numpy as pnp
import numpy as np
from pennylane.wires import Wires

wires_pnp =  Wires(pnp.array([1, 2], requires_grad=True))
wires_np =  Wires(np.array([1, 2]))

print(wires_pnp) # <Wires = [tensor(1, requires_grad=True), tensor(2, requires_grad=True)]>
print(wires_np) # <Wires = [1, 2]>
```
The normal numpy produces the desired behaviour, because when we convert an array to the tuple representing the wire labels, the elements of the array are used. The PennyLane array class in contrast (and for good reasons discussed elsewhere) returns a tuple of 0-dimensional tensors instead. 

Now it gets interesting: this does not seem to cause an issue with circuit execution like
``` python
import pennylane as qml
qml.enable_tape()

N_states = 4

def circuit(wires):
    qml.RY(0.1, wires=wires[0]) # this gets converted internally to <Wires=[0]> because 0-dim iterables are unpacked
    qml.CRY(0.1, wires=[wires[0], wires[2]]) # this gets converted to <Wires=[tensor([0], requires_grad=True),..]>
    return qml.expval(qml.PauliZ(0)) # this integer gets converted to <Wires=[0]>

dev = qml.device("default.qubit", wires=[0, 1, 2, 3])
my_qnode = qml.QNode(circuit, dev) # qnode chooses autograd device which takes items out of tensor-valued wire labels

print(my_qnode(range(4)))
```
because the interface default devices have the line 
`wires = Wires([w.item() if isinstance(w, tensor) else w for w in wires])`, so they unpack tensor wire labels and make the logic that should fail pass. However, the circuit drawer does not do unpacking, so it gets confused and treats the tensor wires as different objects from the non-tensor ones. 

If it wasn't for that sneaky unpacking line, this ugly but completely logical workflow passes fine: 

``` python
import pennylane as qml
import pennylane.numpy as pnp
qml.enable_tape()

N_states = 4

def circuit(wires):
    qml.RY(0.1, wires=[wires[0]])
    qml.CRY(0.1, wires=[wires[0], wires[2]])
    return qml.expval(qml.PauliZ([wires[0]]))

dev = qml.device("default.qubit", wires=[pnp.array(i) for i in range(4)])
my_qnode = qml.QNode(circuit, dev)

print(my_qnode(range(4)))
print(my_qnode.qtape.operations)
print(my_qnode.draw(wire_order=[pnp.array(i) for i in range(4)]))
```
Note that this is the desired logic we want to enforce: since the wires are cast into tensors, the user needs to address them with tensors.

All in all I would recommend the following: 
1. We remove the line that unpacks tensor wires
2. When users want to pass wires to circuits we inform them about the backwards compatibility fix and its consequences and advise them to use tensors as wire labels 
3. As soon as possible we remove the internal conversion, and everything should follow the expected logic.

Let me know what you guys think and I can make the changes.

**Comment by mariaschuld** [comment]
This should work now:

```python
import pennylane as qml
from pennylane import numpy as pnp
import numpy as np
qml.enable_tape()
N_states = 4
params = np.random.uniform(low=-np.pi/2, high=np.pi/2, size=(N_states))
def circuit(params, wires):
    qml.broadcast(qml.RY, wires, pattern='single', parameters=params)
    return qml.expval(qml.PauliZ(pnp.array(0)))

dev = qml.device("default.qubit", wires=[pnp.array(i) for i in range(N_states)])

my_qnode = qml.QNode(circuit, dev)
my_qnode(params, range(N_states))
print(my_qnode.draw())
```

In future, after we broke with backwards compatibility that turns positional qnode arguments to tensors, one will be able to address wires as usual, so this cumbersome notation is only temporary.
